# S06_T01: Tasca mètodes de mostreig

## Descripció:
Aprèn a realitzar mostreig de les dades amb Python.

In [2]:
#Importamos las librerias:
import numpy as np
import pandas as pd
import random

pd.set_option('display.max_columns', None)


## Nivell 1
### Exercici 1:
Agafa un conjunt de dades de tema esportiu que t'agradi. Realitza un mostreig de les dades generant una mostra aleatòria simple i una mostra sistemàtica.

In [3]:
# Utilizaremos un dataset de jugadores de la NBA: https://www.kaggle.com/justinas/nba-players-data
df = pd.read_csv(r'..\DB\all_seasons.csv')
df.head()

Unnamed: 0        player_name team_abbreviation   age  player_height  \
0           0      Dennis Rodman               CHI  36.0         198.12   
1           1  Dwayne Schintzius               LAC  28.0         215.90   
2           2       Earl Cureton               TOR  39.0         205.74   
3           3        Ed O'Bannon               DAL  24.0         203.20   
4           4        Ed Pinckney               MIA  34.0         205.74   

   player_weight                      college country draft_year draft_round  \
0      99.790240  Southeastern Oklahoma State     USA       1986           2   
1     117.933920                      Florida     USA       1990           1   
2      95.254320                Detroit Mercy     USA       1979           3   
3     100.697424                         UCLA     USA       1995           1   
4     108.862080                    Villanova     USA       1985           1   

  draft_number  gp  pts   reb  ast  net_rating  oreb_pct  dreb_pct  usg_pct  \
0           27  55  5.7  16.1  3.1        16.1     0.186     0.323    0.100   
1           24  15  2.3   1.5  0.3        12.3     0.078     0.151    0.175   
2           58   9  0.8   1.0  0.4        -2.1     0.105     0.102    0.103   
3            9  64  3.7   2.3  0.6        -8.7     0.060     0.149    0.167   
4           10  27  2.4   2.4  0.2       -11.2     0.109     0.179    0.127   

   ts_pct  ast_pct   season  
0   0.479    0.113  1996-97  
1   0.430    0.048  1996-97  
2   0.376    0.148  1996-97  
3   0.399    0.077  1996-97  
4   0.611    0.040  1996-97

In [ ]:
df.describe()

Unnamed: 0           age  player_height  player_weight            gp  \
count  11145.000000  11145.000000   11145.000000   11145.000000  11145.000000   
mean    5572.000000     27.168686     200.812818     100.637868     52.005832   
std     3217.428709      4.344164       9.190973      12.576295     25.069495   
min        0.000000     18.000000     160.020000      60.327736      1.000000   
25%     2786.000000     24.000000     195.580000      90.718400     32.000000   
50%     5572.000000     27.000000     200.660000      99.790240     58.000000   
75%     8358.000000     30.000000     208.280000     109.315672     74.000000   
max    11144.000000     44.000000     231.140000     163.293120     85.000000   

                pts           reb           ast    net_rating      oreb_pct  \
count  11145.000000  11145.000000  11145.000000  11145.000000  11145.000000   
mean       8.126487      3.560036      1.801463     -2.153899      0.055593   
std        5.935482      2.495394      1.789940     12.150611      0.043889   
min        0.000000      0.000000      0.000000   -200.000000      0.000000   
25%        3.500000      1.800000      0.600000     -6.300000      0.022000   
50%        6.600000      3.000000      1.200000     -1.300000      0.043000   
75%       11.500000      4.700000      2.400000      3.200000      0.086000   
max       36.100000     16.300000     11.700000    300.000000      1.000000   

           dreb_pct       usg_pct        ts_pct       ast_pct  
count  11145.000000  11145.000000  11145.000000  11145.000000  
mean       0.141772      0.185599      0.508099      0.131078  
std        0.063194      0.053047      0.098879      0.095017  
min        0.000000      0.000000      0.000000      0.000000  
25%        0.096000      0.150000      0.478000      0.065000  
50%        0.132000      0.182000      0.521000      0.102000  
75%        0.182000      0.218000      0.557000      0.178000  
max        1.000000      1.000000      1.500000      1.000000

Observamos que tenemos 11145 registros, por lo que para realizar las diferentes muestras, escogemos un valor n de un 10% aprox: n = 1000. Por lo que cada muestra que creemos será de 1000 "individuos".

In [ ]:
# Muestra aleatoria simple:
n = 1000
df_simple = df.sample(n=n)
df_simple.head()

Unnamed: 0       player_name team_abbreviation   age  player_height  \
1565         1565    Ryan Robertson               SAC  23.0         195.58   
7742         7742     Alexis Ajinca               NOP  26.0         218.44   
7661         7661  Samuel Dalembert               DAL  34.0         210.82   
10462       10462   Thabo Sefolosha               UTA  35.0         200.66   
7994         7994        Mike James               CHI  39.0         187.96   

       player_weight     college      country draft_year draft_round  \
1565       86.182480      Kansas          USA       1999           2   
7742      112.490816        None       France       2008           1   
7661      113.398000  Seton Hall        Haiti       2001           1   
10462      99.790240        None  Switzerland       2006           1   
7994       85.275296    Duquesne          USA  Undrafted   Undrafted   

      draft_number  gp  pts  reb  ast  net_rating  oreb_pct  dreb_pct  \
1565            45   1  5.0  0.0  0.0       -33.4     0.000     0.000   
7742            20  56  5.9  4.9  0.7        -5.7     0.120     0.222   
7661            26  80  6.6  6.8  0.5         1.8     0.146     0.242   
10462           13  50  3.8  2.5  0.5         8.8     0.014     0.181   
7994     Undrafted  11  1.0  0.6  1.5        -5.1     0.000     0.121   

       usg_pct  ts_pct  ast_pct   season  
1565     0.119   0.388    0.000  1999-00  
7742     0.166   0.589    0.067  2013-14  
7661     0.150   0.605    0.035  2013-14  
10462    0.129   0.599    0.059  2018-19  
7994     0.175   0.251    0.347  2013-14

In [ ]:
# Muestra aleatoria sistemática:
# Obtenemos el primer valor aleatorio (primero), y el valor de cada "salto"
primero = random.randint(0,100)
salto = round(len(df)/n)
print(f'El primer registro será el número {primero}, y se escogerán {n} registros con un intervalo de {salto}')

El primer registro será el número 49, y se escogerán 1000 registros con un intervalo de 11


In [ ]:
# Y generamos un nuevo df con una muestra sistemática
indices = np.arange(primero, len(df), salto)
print(len(indices))
df_sistematica = df.iloc[indices]
df_sistematica.head()

1009


Unnamed: 0         player_name team_abbreviation   age  player_height  \
49          49  Dontonio Wingfield               POR  23.0         203.20   
60          60          Ike Austin               MIA  27.0         208.28   
71          71        Jason Caffey               CHI  24.0         203.20   
82          82         Henry James               ATL  31.0         203.20   
93          93           Glen Rice               CHH  30.0         203.20   

    player_weight          college country draft_year draft_round  \
49     116.119552       Cincinnati     USA       1994           2   
60     122.469840    Arizona State     USA       1991           2   
71     116.119552          Alabama     USA       1995           1   
82      99.790240  St. Mary's (TX)     USA  Undrafted   Undrafted   
93      99.790240         Michigan     USA       1989           1   

   draft_number  gp   pts  reb  ast  net_rating  oreb_pct  dreb_pct  usg_pct  \
49           37  47   4.5  2.9  1.0         1.3     0.132     0.144    0.204   
60           48  82   9.7  5.8  1.2         2.9     0.090     0.218    0.227   
71           20  75   7.3  4.0  1.2         9.0     0.114     0.134    0.184   
82    Undrafted  53   6.7  1.5  0.4         1.2     0.034     0.067    0.171   
93            4  79  26.8  4.0  2.0         3.2     0.025     0.087    0.272   

    ts_pct  ast_pct   season  
49   0.501    0.140  1996-97  
60   0.536    0.100  1996-97  
71   0.574    0.094  1996-97  
82   0.555    0.036  1996-97  
93   0.605    0.088  1996-97

In [ ]:
'''
# Define systematic sampling function
def systematic_sampling(df, step):
    
    indexes = np.arange(0,len(df),step=step)
    systematic_sample = df.iloc[indexes]
    return systematic_sample
    
# Obtain a systematic sample and save it in a new variable
systematic_sample = systematic_sampling(df, salto)

# View sampled data frame
systematic_sample
'''

'\n# Define systematic sampling function\ndef systematic_sampling(df, step):\n    \n    indexes = np.arange(0,len(df),step=step)\n    systematic_sample = df.iloc[indexes]\n    return systematic_sample\n    \n# Obtain a systematic sample and save it in a new variable\nsystematic_sample = systematic_sampling(df, salto)\n\n# View sampled data frame\nsystematic_sample\n'

## Nivell 2
### Exercici 2:
Continua amb el conjunt de dades de tema esportiu i genera una mostra estratificada i una mostra utilitzant SMOTE (Synthetic Minority Oversampling Technique).

In [ ]:
normalizacion_antes = df.team_abbreviation.value_counts(normalize=True).sort_index()
# Vemos que puede ser interesante generar una muestra estratificada por equipos

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X_train, X_test = train_test_split(df, test_size=0.1, stratify=df[['team_abbreviation']])
print(f'X_test shape: {X_test.shape}, X_train shape: {X_train.shape}\n')
normalizacion_despues = X_test.team_abbreviation.value_counts(normalize=True).sort_index()


X_test shape: (1115, 22), X_train shape: (10030, 22)



Vemos que la forma de la muestra estratificada es del 10% de la población total. A continuación comprobaremos si todos los equipos aparecen correctamente representados:

In [ ]:
comparativa_normalizacion = pd.concat([normalizacion_antes,normalizacion_despues], axis = 1)
comparativa_normalizacion.columns = [['Antes de estratificar','Después de estratificar']]
comparativa_normalizacion['Diferencia'] = normalizacion_antes - normalizacion_despues
comparativa_normalizacion

Antes de estratificar Después de estratificar Diferencia
ATL              0.034365                0.034081   0.000284
BKN              0.012562                0.012556   0.000006
BOS              0.032930                0.033184  -0.000254
CHA              0.022790                0.022422   0.000369
CHH              0.007986                0.008072  -0.000086
CHI              0.033109                0.033184  -0.000075
CLE              0.034993                0.034978   0.000016
DAL              0.034455                0.034081   0.000374
DEN              0.033378                0.033184   0.000194
DET              0.032122                0.032287  -0.000165
GSW              0.033737                0.034081  -0.000344
HOU              0.033917                0.034081  -0.000164
IND              0.033288                0.033184   0.000105
LAC              0.034904                0.034978  -0.000074
LAL              0.033288                0.033184   0.000105
MEM              0.028264                0.028700  -0.000436
MIA              0.034724                0.034978  -0.000253
MIL              0.033288                0.033184   0.000105
MIN              0.032840                0.033184  -0.000344
NJN              0.023060                0.023318  -0.000259
NOH              0.012831                0.012556   0.000275
NOK              0.002871                0.002691   0.000181
NOP              0.010947                0.010762   0.000184
NYK              0.033288                0.033184   0.000105
OKC              0.017497                0.017040   0.000456
ORL              0.033019                0.033184  -0.000165
PHI              0.034096                0.034081   0.000015
PHX              0.032481                0.032287   0.000194
POR              0.032930                0.033184  -0.000254
SAC              0.032122                0.032287  -0.000165
SAS              0.033827                0.034081  -0.000254
SEA              0.016330                0.016143   0.000187
TOR              0.034993                0.034978   0.000016
UTA              0.032301                0.032287   0.000014
VAN              0.006460                0.006278   0.000182
WAS              0.034006                0.034081  -0.000074

Podemos observar que la muestra estratificada conserva correctamente los porcentajes de cada equipo en la población inicial: las variaciones son menores a un 0,00%

A continuación, extraeremos una muestra utilizando el método SMOTE:

In [ ]:
# Hemos instalado el paquete imbalanced-learn con: conda install -c conda-forge imbalanced-learn
# import imblearn #lo importamos
from imblearn.over_sampling import SMOTE

In [ ]:
# Para poder utilizar este método, nos centraremos en si los jugadores son estadounidenses (0) o extranjeros (1)
# muy probablemente la gran mayoria de jugadores sean americanos, por lo que los extranjeros contarán con una representación menor,
# por lo que según que tipo de muestreo que se realice, podría tener una baja representación
usa_counts_before = df.loc[df.country == 'USA', 'country'].count()
df2 = df.copy()
print(f'USA counts antes: {usa_counts_before}')
df2['country'] = df['country'].apply(lambda x: 0 if x == 'USA' else 1)

usa_counts_after = df2.loc[df2.country == 0, 'country'].count()
print(f'USA counts después: {usa_counts_after}')

USA counts antes: 9410
USA counts después: 9410


In [ ]:
# Contamos cuantos 0 (USA) y 1 (otros) hay:
df2.country.value_counts()

0    9410
1    1735
Name: country, dtype: int64

In [ ]:
# Eliminamos las columnas con str:
columns = ['player_name', 'team_abbreviation', 'college', 'draft_year', 'draft_round', 'draft_number', 'season']
df2 = df2.drop(columns=columns)

In [ ]:
# Con SMOTE, queremos igualar esos 1 a la cantidad de 0 que hay
oversample = SMOTE()
X, y = oversample.fit_resample(df2, df2[['country']])
X.country.value_counts()

0    9410
1    9410
Name: country, dtype: int64

In [ ]:
df2.describe() # Done!! :D

Unnamed: 0           age  player_height  player_weight       country  \
count  11145.000000  11145.000000   11145.000000   11145.000000  11145.000000   
mean    5572.000000     27.168686     200.812818     100.637868      0.155675   
std     3217.428709      4.344164       9.190973      12.576295      0.362563   
min        0.000000     18.000000     160.020000      60.327736      0.000000   
25%     2786.000000     24.000000     195.580000      90.718400      0.000000   
50%     5572.000000     27.000000     200.660000      99.790240      0.000000   
75%     8358.000000     30.000000     208.280000     109.315672      0.000000   
max    11144.000000     44.000000     231.140000     163.293120      1.000000   

                 gp           pts           reb           ast    net_rating  \
count  11145.000000  11145.000000  11145.000000  11145.000000  11145.000000   
mean      52.005832      8.126487      3.560036      1.801463     -2.153899   
std       25.069495      5.935482      2.495394      1.789940     12.150611   
min        1.000000      0.000000      0.000000      0.000000   -200.000000   
25%       32.000000      3.500000      1.800000      0.600000     -6.300000   
50%       58.000000      6.600000      3.000000      1.200000     -1.300000   
75%       74.000000     11.500000      4.700000      2.400000      3.200000   
max       85.000000     36.100000     16.300000     11.700000    300.000000   

           oreb_pct      dreb_pct       usg_pct        ts_pct       ast_pct  
count  11145.000000  11145.000000  11145.000000  11145.000000  11145.000000  
mean       0.055593      0.141772      0.185599      0.508099      0.131078  
std        0.043889      0.063194      0.053047      0.098879      0.095017  
min        0.000000      0.000000      0.000000      0.000000      0.000000  
25%        0.022000      0.096000      0.150000      0.478000      0.065000  
50%        0.043000      0.132000      0.182000      0.521000      0.102000  
75%        0.086000      0.182000      0.218000      0.557000      0.178000  
max        1.000000      1.000000      1.000000      1.500000      1.000000

## Nivell 3
### Exercici 3:
Continua amb el conjunt de dades de tema esportiu i genera una mostra utilitzant el mètode Reservoir sampling.

Reservoir sampling: se trata de un algoritmo para escoger aleatoriamente k muestras de una lista de n ítems, donde n suele ser un número muy grande o desconocido. El algoritmo solo pasa una vez por los n ítems y no "mira atrás" en la lista. 

In [ ]:
import random # importamos random para generar un número aleatorio

In [ ]:
# en nuestro caso, n = len(df) y k escogeremos 100
n = len(df)
k = 100
print(f'El tamaño de la poblaclión es {n} y el tamaño de la muestra será {k}')

El tamaño de la poblaclión es 11145 y el tamaño de la muestra será 100


In [ ]:
# Creamos una lista vacia en la que se irán introduciendo los ítems de la muestra
# código adaptado del de la siguiente url: https://towardsdatascience.com/coding-interview-questions-reservoir-sampling-687ff5b6822b
reservorio = []
i = 0
for item in df.index:               # iteramos en los índices del df (población)
    i = i + 1 
    if len(reservorio) < k:         # si la lista reservorio tiene menos valores que los que queremos de muestra
        reservorio.append(item)     # añadimos el item a la lista reservorio
    else:                           
        j = int(random.random() * n) # generamos números aleatorios 
        if j < k:                   # si el número aleatorio es menor que k, 
            reservorio[j] = item    # substituimos el item en la posición aleatoria j
reservorio

[0,
 10163,
 6829,
 7192,
 9155,
 5,
 6,
 7,
 4124,
 9,
 10307,
 11,
 12,
 3403,
 6561,
 1118,
 16,
 6466,
 9969,
 19,
 20,
 21,
 3125,
 148,
 24,
 8618,
 7260,
 11010,
 6689,
 10929,
 10247,
 1586,
 6062,
 33,
 9668,
 933,
 7900,
 10364,
 9491,
 8083,
 40,
 6339,
 5164,
 8804,
 44,
 9542,
 5892,
 47,
 7337,
 49,
 5817,
 51,
 52,
 10381,
 54,
 6846,
 6855,
 57,
 2324,
 5575,
 5959,
 5408,
 62,
 3896,
 938,
 65,
 66,
 10171,
 4741,
 10343,
 6866,
 4674,
 2237,
 73,
 74,
 75,
 9081,
 2033,
 78,
 79,
 6828,
 10602,
 1972,
 9679,
 4061,
 10455,
 3891,
 6578,
 88,
 89,
 90,
 91,
 8529,
 3802,
 410,
 1472,
 96,
 6870,
 98,
 4862]

In [ ]:
type(reservorio)

list

In [ ]:
# Creamos la muestra con los indices obtenidos mediante reservoir sampling
df_reservoir = df.iloc[reservorio]
df_reservoir

Unnamed: 0        player_name team_abbreviation   age  player_height  \
0               0      Dennis Rodman               CHI  36.0         198.12   
10163       10163     Jemerrio Jones               LAL  24.0         195.58   
6829         6829    Mickael Pietrus               BOS  30.0         198.12   
7192         7192         Will Bynum               DET  30.0         182.88   
9155         9155          Ty Lawson               SAC  29.0         180.34   
...           ...                ...               ...   ...            ...   
1472         1472  Anfernee Hardaway               PHX  28.0         200.66   
96             96         Grant Long               DET  31.0         205.74   
6870         6870      Shannon Brown               PHX  26.0         193.04   
98             98       Greg Anthony               VAN  29.0         185.42   
4862         4862        Mark Madsen               MIN  31.0         205.74   

       player_weight                      college country draft_year  \
0          99.790240  Southeastern Oklahoma State     USA       1986   
10163      78.925008                         None     USA  Undrafted   
6829       97.522280                         None  France       2003   
7192       83.914520                 Georgia Tech     USA  Undrafted   
9155       88.450440               North Carolina     USA       2009   
...              ...                          ...     ...        ...   
1472       97.522280                      Memphis     USA       1993   
96        112.490816             Eastern Michigan     USA       1988   
6870       95.254320               Michigan State     USA       2006   
98         81.646560             Nevada-Las Vegas     USA       1991   
4862      113.398000                     Stanford     USA       2000   

      draft_round draft_number  gp   pts   reb  ast  net_rating  oreb_pct  \
0               2           27  55   5.7  16.1  3.1        16.1     0.186   
10163   Undrafted    Undrafted   6   4.5   8.2  2.2        10.7     0.100   
6829            1           11  42   6.9   3.1  0.6         2.5     0.024   
7192    Undrafted    Undrafted  65   9.8   1.5  3.6        -6.6     0.026   
9155            1           18  69   9.9   2.6  4.8        -4.2     0.029   
...           ...          ...  ..   ...   ...  ...         ...       ...   
1472            1            3  60  16.9   5.8  5.3         8.8     0.047   
96              2           33  65   5.0   3.4  0.6         4.0     0.096   
6870            1           25  59  11.0   2.7  1.2        -2.0     0.030   
98              1           12  65   9.5   2.8  6.3        -9.4     0.015   
4862            1           29  56   1.1   1.6  0.2         0.4     0.096   

       dreb_pct  usg_pct  ts_pct  ast_pct   season  
0         0.323    0.100   0.479    0.113  1996-97  
10163     0.198    0.108   0.398    0.115  2018-19  
6829      0.135    0.162   0.509    0.045  2011-12  
7192      0.068    0.267   0.532    0.341  2012-13  
9155      0.087    0.193   0.551    0.300  2016-17  
...         ...      ...     ...      ...      ...  
1472      0.121    0.212   0.549    0.230  1999-00  
96        0.150    0.154   0.523    0.058  1996-97  
6870      0.094    0.228   0.507    0.083  2011-12  
98        0.099    0.177   0.526    0.358  1996-97  
4862      0.129    0.076   0.547    0.031  2006-07  

[100 rows x 22 columns]